endstate_correction/smc.py

```diff
@@ -0,0 +1,146 @@
+        # initialize weights to be uniform
+        # select random subset of walkers from the samples
+        # for each interpolation step i between \lamb=0 and \lamb=1
+        #   propagate the walkers for a fixed number of steps with u(x; \lamb=i) to generate pool of IID samples
+        #   for each walker j
+        #       calculate the unnormalized weights w = exp(u(x; \lamb=i+1) - u(x; \lamb=i))
+        #   resample the walkers according to the normalized weights
+
+        # calculate the free energy estimate using the weights of each walker and the Zwanzig relation
+
+        # initialize weights
+        weights = np.ones(nr_of_walkers) / nr_of_walkers
+        # initialize lambda values
+        lambdas = np.linspace(0, 1, nr_of_steps)
+
+        # select initial samples
+        random_frame_idxs = np.random.choice(
+            len(self.samples.xyz) - 1, size=nr_of_walkers
+        )
+        walkers = [
+            self.samples.openmm_positions(random_frame_idx)
+            for random_frame_idx in random_frame_idxs
+        ]
+
+        assert len(walkers) == nr_of_walkers
+
+        # start with switch
+        for lamb_idx in tqdm(range(len(lambdas) - 1)):
+            # set lambda parameter
+            self.sim.context.setParameter("lambda_interpolate", lambdas[lamb_idx])
+            # Propagate the walkers
+            walkers = self._propagate_particles(walkers, self.sim)
+            # calculate work
+            # evaluate U_(\lamb_(i+1))(x_i) -  U_(\lamb_(i))(x_i)
+            # calculate U_(\lamb_(i))(x_i)
+            u_now = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx], walkers, self.sim
+            )
+            # calculate U_(\lamb_(i+1))(x_i)
+            u_future = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx + 1], walkers, self.sim
+            )
+            # calculate weights (equation 2 in 10.1021/acs.jctc.1c01198)
+            current_deltaEs = u_future - u_now  # calculate difference in energy
+            current_weights = np.exp(
+                np.nanmin(current_deltaEs) - current_deltaEs
+            )  # subtract minimum to avoid numerical issues
+            current_weights /= np.sum(current_weights)  # normalize
+            # calculate ESS
+            ESS = 1.0 / np.sum(current_weights**2)
+            print(f"Effective Sample Size at lambda = {lambdas[lamb_idx]}: {ESS}")
+            print(current_weights)
+            # add to accumulated logZ
+            self.logZ += logsumexp(-current_deltaEs) - np.log(nr_of_walkers)
+
+            # Resample the particles based on the weights
+            random_frame_idxs = np.random.choice(
+                nr_of_walkers, size=nr_of_walkers, p=current_weights
+            )
+            walkers = [walkers[idx] for idx in random_frame_idxs]
+
+        # reset lambda value
+        self.sim.context.setParameter("lambda_interpolate", 0.0)
```

Comment:
Do you want to do that though? It's an irreversible method so good to keep it at 1.